### PR TITLE
evaluator: impl json_object and json_array

### PR DESCRIPTION
**Pull request by lishihai9017**
Hi
This PR add JSON_OBECT and JSON_ARRAY function.

PTAL @AndreMouche @andelf @hhkbp2

### CODE REVIEW

File: src/util/codec/mysql/json/json_array.rs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+// FIXME: remove following later
+#![allow(dead_code)]
+
+use super::Json;
+use super::super::super::datum::Datum;
+use super::super::super::Result;
+
+impl Json {
+    // https://dev.mysql.com/doc/refman/5.7/en/json-creation-functions.html#function_json-array
+    pub fn to_array(elems: &[Datum]) -> Result<Json> {
```
Comment (AndreMouche) [inline src/util/codec/mysql/json/json_array.rs]: Could we use Vec<Datum> directly here?

Comment (AndreMouche) [inline src/util/codec/mysql/json/json_array.rs]: I prefer to implement this function in `evaluator` with the following reasons:
1. we never use `Datum` in this mod
2. keep consistent with `tidb`

File: src/util/codec/mysql/json/json_object.rs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        let mut map = BTreeMap::new();
+        let mut idx = 0;
+        while idx < len {
+            let key = match kvs[idx] {
+                Datum::Null => {
+                    return Err(invalid_type!("JSON documents may not contain NULL member names"));
+                }
+                _ => try!(kvs[idx].to_string()),
+            };
+            let val = try!(kvs[idx + 1].clone().into_json());
```
Comment (AndreMouche) [inline src/util/codec/mysql/json/json_object.rs]: `clone` is not a good habit

File: src/util/codec/mysql/json/mod.rs [hunk 3]
```diff
@@ -10,8 +10,10 @@
 // distributed under the License is distributed on an "AS IS" BASIS,
 // See the License for the specific language governing permissions and
 // limitations under the License.
+
 // FIXME(shirly): remove following later
 #![allow(dead_code)]
```
Comment (AndreMouche) [inline src/util/codec/mysql/json/mod.rs]: Could we remove this line now?

File: src/util/xeval/evaluator.rs [hunk 4]
```diff
@@ -513,6 +515,18 @@ impl Evaluator {
...
 
+    fn eval_json_object(&mut self, ctx: &EvalContext, expr: &Expr) -> Result<Datum> {
+        let children = try!(self.eval_more_children(ctx, expr, 0));
+        let obj = try!(Json::to_object(&children));
+        Ok(Datum::Json(obj))
+    }
+
+    fn eval_json_array(&mut self, ctx: &EvalContext, expr: &Expr) -> Result<Datum> {
+        let children = try!(self.eval_more_children(ctx, expr, 0));
+        let arr = try!(Json::to_array(&children));
```
Comment (AndreMouche) [inline src/util/xeval/evaluator.rs]: I'd like to implement `to_array` in this file.

File: src/util/codec/mysql/json/json_array.rs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        let mut arr = Vec::with_capacity(elems.len());
```
Comment (hhkbp2) [inline src/util/codec/mysql/json/json_array.rs]: s/arr/a/
or
s/arr/array/

File: src/util/codec/mysql/json/json_array.rs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_json_array() {
+        let d = vec![Datum::I64(1),
+                     Datum::Bytes(b"sdf".to_vec()),
+                     Datum::U64(2),
+                     Datum::Json(r#"[3,4]"#.parse().unwrap())];
+        let ep_json = r#"[1,"sdf",2,[3,4]]"#.parse().unwrap();
```
Comment (hhkbp2) [inline src/util/codec/mysql/json/json_array.rs]: what is `ep`?

File: src/util/codec/mysql/json/json_array.rs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        assert_eq!(Json::to_array(&d).unwrap(), ep_json);
+
+        let d = vec![];
+        let ep_json = "[]".parse().unwrap();
+        assert_eq!(Json::to_array(&d).unwrap(), ep_json);
```
Comment (hhkbp2) [inline src/util/codec/mysql/json/json_array.rs]: Use this form:
```
let mut test_cases = vec![(datum_vec, expected_json), ...];
for (i, (v, j)) in test_cases.drain(..).enumerate() {
     let r = to_array()
     assert!(r.is_ok());
     ......
     assert_eq!(...);
}
```
to remove duplicated code.

File: src/util/codec/mysql/json/json_array.rs (hunk 6 above)
Comment (lishihai9017) [inline src/util/codec/mysql/json/json_array.rs]: expect

File: src/util/codec/mysql/json/json_object.rs [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+use super::Json;
+use super::super::super::datum::Datum;
+use super::super::super::{Result, Error};
+
+impl Json {
+    // https://dev.mysql.com/doc/refman/5.7/en/json-creation-functions.html#function_json-object
+    pub fn to_object(kvs: &[Datum]) -> Result<Json> {
+        let len = kvs.len();
+        if len & 1 == 1 {
```
Comment (hhkbp2) [inline src/util/codec/mysql/json/json_object.rs]: Add an util function `is_odd()` for this line.

File: src/util/codec/mysql/json/json_object.rs [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        assert!(Json::to_object(&d).is_err());
+
+        let d = vec![Datum::I64(1),
+                     Datum::Bytes(b"sdf".to_vec()),
+                     Datum::Bytes(b"asd".to_vec()),
+                     Datum::Bytes(b"qwe".to_vec()),
+                     Datum::I64(2),
+                     Datum::Json(r#"{"3":4}"#.parse().unwrap())];
+        let ep_json = r#"{"1":"sdf","2":{"3":4},"asd":"qwe"}"#.parse().unwrap();
+        assert_eq!(Json::to_object(&d).unwrap(), ep_json);
```
Comment (hhkbp2) [inline src/util/codec/mysql/json/json_object.rs]: Dito. List the test cases to avoid duplicate code.

**Comment by hhkbp2** [comment]
Rest LGTM

File: src/util/codec/mysql/json/json_array.rs (hunk 1 above)
Comment (lishihai9017) [inline src/util/codec/mysql/json/json_array.rs]: i think we should not mix evaluator and codec logic

File: src/util/codec/mysql/json/json_array.rs [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+use super::super::super::Result;
+
+impl Json {
+    // https://dev.mysql.com/doc/refman/5.7/en/json-creation-functions.html#function_json-array
+    pub fn to_array(elems: &[Datum]) -> Result<Json> {
+        let mut arr = Vec::with_capacity(elems.len());
+        for elem in elems {
+            arr.push(try!(elem.clone().into_json()));
+        }
+        Ok(Json::Array(arr))
```
Comment (andelf) [inline src/util/codec/mysql/json/json_array.rs]: elems.iter().cloned().map(|e| e.into_json()).collect().map(Json::Array)

🤣

File: src/util/codec/mysql/json/json_object.rs (hunk 2 above)
Comment (andelf) [inline src/util/codec/mysql/json/json_object.rs]: will overflows

Comment (andelf) [inline src/util/codec/mysql/json/json_object.rs]: should we introduce  https://crates.io/crates/itertools

File: src/util/codec/mysql/json/json_array.rs (hunk 10 above)
Comment (lishihai9017) [inline src/util/codec/mysql/json/json_array.rs]: 666

File: src/util/codec/mysql/json/mod.rs (hunk 3 above)
Comment (lishihai9017) [inline src/util/codec/mysql/json/mod.rs]: warning at PRECEDENCE_*

File: src/util/codec/mysql/json/mod.rs [hunk 11]
```diff
@@ -47,3 +53,77 @@ pub enum Json {
...
+    if !is_even(len) {
+        return Err(Error::Other(box_err!("Incorrect parameter count in the call to native \
+                                          function 'JSON_OBJECT'")));
+    }
+    let mut map = BTreeMap::new();
+    let mut key = None;
+    for elem in kvs.into_iter() {
+        if key.is_none() {
+            // take elem as key
+            if elem == Datum::Null {
```
Comment (AndreMouche) [inline src/util/codec/mysql/json/mod.rs]: Could you check this key first like `eval_json_modify` did?

File: src/util/codec/mysql/json/mod.rs [hunk 12]
```diff
@@ -47,3 +53,77 @@ pub enum Json {
...
+                return Err(invalid_type!("JSON documents may not contain NULL member names"));
+            }
+            key = Some(try!(elem.into_string()));
+        } else {
```
Comment (AndreMouche) [inline src/util/codec/mysql/json/mod.rs]: I'd like to get `<key,value>` in pair instead of one by one.

**Comment by AndreMouche** [comment]
Please resolve conflicts to offer the reviewers convenience

File: src/util/codec/mysql/json/mod.rs (hunk 12 above)
Comment (hicqu) [inline src/util/codec/mysql/json/mod.rs]: you can use `let keys, vals =kvs.into_iter().enumerate().partition(|(&n, &m)| n & 1 == 0)`

**Comment by hhkbp2** [comment]
Any update? @lishihai9017

File: src/util/codec/mysql/json/mod.rs (hunk 12 above)
Comment (lishihai9017) [inline src/util/codec/mysql/json/mod.rs]: the trait `std::iter::Extend<(_, _)>` is not implemented for `(_, _)`

**Pull request merged**